### ISSUE TITLE
Deduplicate push subscriptions by subscription endpoint (current behavior causes severe performance degradation)

### ISSUE THREAD

**Issue by inga-lovinde** [mastodon/mastodon#30544]
### Pitch

Currently there might be several push subscriptions for the same endpoint (probably e.g. if an user logged in into the same client app multiple times without cleanly logging out). And a separate task will be created for each notification for each of these endpoints, overloading the queue.

### Motivation

My instance has a mostly inactive user who somehow has 150 subscriptions for just two endpoints (10 very old ones for one endpoint, 140 newer ones for another).

This means that every push notification for them gets sent 150 times when 2 would suffice.

And this in turn means that if these endpoints are pointing to ntfy (as in https://github.com/mastodon/mastodon/issues/26078 ), then even a handful of notifications schedules thousands of requests to ntfy, which results in ntfy fail2banning server IP address, which results in server retrying these thousands of requests multiple times, which results in complete loss of outgoing federation (because it uses the same queue, and sidekiq can only process five tasks at a time, which with 5-second timeout means one task per second, at best).

**Comment by trwnh** [comment]
Maybe related to #30103 ?

**Comment by ThisIsMissEm** [comment]
> Maybe related to #30103 ?

Yeah, they definitely seem related. We could also expire the push subscriptions after a period of time without breaking anything, I think?

**Comment by renchap** [comment]
Yes this is related. The source issue should have been fixed by the linked PR, but existing duplicate subscriptions are not purged. Pinging @clearlyclaire to see if we want to add this in a migration.

**Comment by inga-lovinde** [comment]
@trwnh 
> Maybe related to https://github.com/mastodon/mastodon/issues/30103 ?

I'm not sure what exactly caused 30103, and when would that race condition manifest.

In this specific case, it's a single year-old access token for fedilab, with 140 associated subscriptions, all seemingly identical (except for `key...` fields), all created at very different datetimes over that year.

Yes, would be nice to have a migration that would purge existing duplicate subscriptions at least for the same application_id / endpoint pairs.

**Comment by ThisIsMissEm** [comment]
> Yes this is related. The source issue should have been fixed by the linked PR, but existing duplicate subscriptions are not purged. Pinging @ClearlyClaire to see if we want to add this in a migration.

I was kinda thinking maybe we want a vacuum job here?

**Comment by ClearlyClaire** [comment]
> @trwnh
> 
> > Maybe related to #30103 ?
> 
> I'm not sure what exactly caused 30103, and when would that race condition manifest.

Basically there we two issues:
- concurrent requests would create multiple subscriptions
- creating a new subscription would delete *one* previous subscription, not all of them (since there was supposed to only be one, but could be multiple in practice)

> In this specific case, it's a single year-old access token for fedilab, with 140 associated subscriptions, all seemingly identical (except for `key...` fields), all created at very different datetimes over that year. Another set of 10 subscriptions are for fedilab too, so it must have something to do with how fedilab interacted with mastodon.

The fact that they are created at very different datetimes hints that it may be a different issue then… but also it's possible that many subscriptions were initially made, and they've been slowly replaced by new ones over time.

**Comment by ThisIsMissEm** [comment]
@clearlyclaire did we do the clean up migration to delete duplicate subscriptions?

**Comment by ClearlyClaire** [comment]
No, because we can't know which are the usable ones.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire can't we just take the latest one for a given endpoint?

**Comment by mjankowski** [comment]
> [@ClearlyClaire](https://github.com/ClearlyClaire) can't we just take the latest one for a given endpoint?

LMK if we want a migration to this effect, or if this aspect is a wontfix/cantfix.

**Comment by ClearlyClaire** [comment]
> [@ClearlyClaire](https://github.com/ClearlyClaire) can't we just take the latest one for a given endpoint?

Hm, one thing we can do is deduplicate over `(key_p256dh, key_auth, access_token_id, endpoint)`. I assumed the keys would be different, and thus some of the subscriptions could be functional while others wouldn't be, but this seems unlikely in practice.

If we try to deduplicate beyond that (e.g. on `(access_token_id, endpoint)` or just `endpoint`), we still risk breaking working subscriptions though.

@inga-lovinde do you still have those records in your database? If so, can you check if they share the same `key_p256dh` and `key_auth` values?

**Comment by bcc** [comment]
Raising a hand here as also affected by this - a number of long term fedilab users with multiple entries in web_push_subscriptions - one user having 89 entries for the same access token!

They all seem to have unique values for key_p256dh and key_auth.
